File: Perl/perl_solutions_11-20/Euler_12.pl
use strict;
use warnings;
use Math::Complex;
use Time::HiRes qw( clock );


sub triangle_divisors
{
    my $divisors = 0;
    my $triangle = 1001;
    my $tri = 0;
    foreach my $i (0..1001) {
        $tri += $i;
    }   
    while ($divisors <= 500) {
        foreach my $i (2..int(sqrt($tri)) + 1) {
            if ($tri % $i == 0) {
                $divisors += 1;
            }
        }
        if ($divisors * 2 <= 500) {
            $triangle += 1;
            $tri += $triangle;
            $divisors = 0;
        }
    }
    return $tri;
}

my $start = clock();
my $answer = triangle_divisors();
my $stop = clock();
print "Answer: $answer\n";
printf "Time: %.5f\n", $stop - $start;

File: Python/py_solutions_11-20/Euler_12.py
# What is the value of the first triangle number to have over five-hundred divisors?

from __future__ import print_function

import math
import timeit
import sys


try:
    range = xrange
except NameError:
    pass

sys.setrecursionlimit(15000)
start = timeit.default_timer()

def factor_gen(tri):
    for i in range(2, int(math.sqrt(tri)) + 1):
        if tri % i == 0:
            yield i

def euler_12(n, tri):
    if sum([2 for i in factor_gen(tri)]) + 2 > 500:
        return tri
    return euler_12(n + 1, sum([tri, n, 1]))

print("Answer: {}".format(euler_12(1001, sum(range(1002)))))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))
